### PR TITLE
dm(engine): cache finished status

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: ref #5504 

### What is changed and how it works?
when task is finished, cache the finished status so user can use query the status of the job.
- cache finished status when worker is finished
- when master failover or job updated, new worker will be created and finished status will be reset.

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/jobmaster/dm/dm_jobmaster.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ type JobMaster struct {
 	// only use when init
 	// it will be outdated if user update job cfg.
 	initJobCfg *config.JobCfg
+	// taskID -> FinishedTaskStatus
+	// worker exists after finished, so we need record the finished status for QueryJobStatus
+	finishedStatus sync.Map
```
Comment (lance6716) [inline engine/jobmaster/dm/dm_jobmaster.go:55]: Worth writing in comment that `finishedStatus` will be lost when jobmaster failover.

And can we use Task.Stage to mark a task is finished? So when failover, only the summary status is lost. And maybe we can also store summary status into Task

File: engine/jobmaster/dm/api.go [hunk 2]
```diff
@@ -170,6 +174,11 @@ func (jm *JobMaster) UpdateJobCfg(ctx context.Context, cfg *config.JobCfg) error
 	if err := jm.checkpointAgent.Update(ctx, cfg); err != nil {
 		return err
 	}
+	// reset finished status, all tasks will be restarted now.
+	jm.finishedStatus.Range(func(key interface{}, value interface{}) bool {
```
Comment (lance6716) [inline engine/jobmaster/dm/api.go:178]: maybe create a new sync.Map

File: engine/jobmaster/dm/dm_jobmaster.go [hunk 3]
```diff
@@ -161,6 +162,7 @@ func (jm *JobMaster) handleOnlineStatus(worker framework.WorkerHandle) error {
 		return err
 	}
 
+	jm.finishedStatus.Delete(taskStatus.Task)
```
Comment (lance6716) [inline engine/jobmaster/dm/dm_jobmaster.go:165]: in which case, will worker become Offine with Finished then become Online?

Comment (GMHDBJD) [reply engine/jobmaster/dm/dm_jobmaster.go:165]: when switch to another unit.

**Comment by GMHDBJD** [comment]
ut failed https://github.com/pingcap/tiflow/issues/6665
/run-verify-ci

File: engine/jobmaster/dm/api.go [hunk 4]
```diff
@@ -98,7 +98,13 @@ func (jm *JobMaster) QueryJobStatus(ctx context.Context, tasks []string) (*JobSt
 					// task finished
 					workerID = workerStatus.ID
 					cfgModRevision = workerStatus.CfgModRevision
-					queryStatusResp = &dmpkg.QueryStatusResponse{Unit: workerStatus.Unit, Stage: metadata.StageFinished}
+					finishedStatus, ok := jm.finishedStatus.Load(taskID)
+					if !ok {
+						queryStatusResp = &dmpkg.QueryStatusResponse{Unit: workerStatus.Unit, Stage: metadata.StageFinished, ErrorMsg: fmt.Sprintf("task %s finished but status not found", taskID)}
```
Comment (lance6716) [inline engine/jobmaster/dm/api.go:103]: “not found” may cause worry, maybe "task %s is finished and status has been deleted"

**Review by lance6716 — APPROVED** [review]

**Review by maxshuang — APPROVED** [review]

**Pull request merged**